source/simulator/lateral_averaging.cc

```diff
@@ -448,17 +250,279 @@ namespace aspect
+            {
+              // get the material properties at each quadrature point if necessary
+              in.reinit(fe_values,
+                        cell,
+                        this->introspection(),
+                        this->get_solution());
+              this->get_material_model().evaluate(in, out);
+            }
+
+          for (unsigned int i = 0; i < n_properties; ++i)
+            (*functors[i])(in, out, fe_values, this->get_solution(), output_values[i]);
+
+          for (unsigned int q = 0; q < n_q_points; ++q)
+            {
+              const double depth = this->get_geometry_model().depth(fe_values.quadrature_point(q));
+              // make sure we are rounding down and never end up with idx==num_slices:
+              const double magic = 1.0-2.0*std::numeric_limits<double>::epsilon();
+              const unsigned int idx = static_cast<unsigned int>(std::floor((depth*n_slices)/max_depth*magic));
+
+              Assert(idx<n_slices, ExcInternalError());
+
+              for (unsigned int i = 0; i < n_properties; ++i)
+                values[i][idx] += output_values[i][q] * fe_values.JxW(q);
+
+              volume[idx] += fe_values.JxW(q);
+            }
+        }
+
+    std::vector<double> volume_all(n_slices);
+    Utilities::MPI::sum(volume, this->get_mpi_communicator(), volume_all);
+
+    bool print_under_res_warning=false;
+    for (unsigned int property=0; property<n_properties; ++property)
+      {
+        std::vector<double> values_all(n_slices);
+        Utilities::MPI::sum(values[property], this->get_mpi_communicator(), values_all);
+
+        for (unsigned int i=0; i<n_slices; ++i)
+          {
+            if (volume_all[i] > 0.0)
+              {
+                values[property][i] = values_all[i] / (static_cast<double>(volume_all[i]));
+              }
+            else
+              {
+                print_under_res_warning = true;
+                // Output nan if no quadrature points in depth block
+                values[property][i] = std::numeric_limits<double>::quiet_NaN();
+              }
+          }
+      }
 
+    if (print_under_res_warning)
+      {
+        this->get_pcout() << std::endl
+                          << "**** Warning: When computing depth averages, there is at least one depth band"
+                          << std::endl
+                          << "     that does not have any quadrature points in it."
+                          << std::endl
+                          << "     Consider reducing the number of depth layers for averaging."
+                          << std::endl << std::endl;
+      }
+  }
+
+  template <int dim>
+  void LateralAveraging<dim>::get_temperature_averages(std::vector<double> &values) const
+  {
+    std::vector<std::vector<double> > vector_values(1,values);
+    std::vector<std::string> property_name(1,"temperature");
+    get_averages(property_name,vector_values);
+    values.swap(vector_values[0]);
```

Comment:
if you return the data from `get_averages` as mentioned earlier, you can just do
```
  values = get_averages (property_name);
```